packages/service-core/src/sync/sync.ts

```diff
@@ -193,27 +205,53 @@ async function* streamResponseInner(
         checkpoint: {
           last_op_id: checkpoint,
           write_checkpoint: writeCheckpoint ? String(writeCheckpoint) : undefined,
-          buckets: [...checksumMap.values()]
+          buckets: [...checksumMap.values()].map((e) => ({
+            ...e,
+            priority: dataBuckets.get(e.bucket)!.description!.priority
+          }))
         }
       };
       yield checksum_line;
     }
     lastChecksums = checksumMap;
     lastWriteCheckpoint = writeCheckpoint;
 
-    // This incrementally updates dataBuckets with each individual bucket position.
-    // At the end of this, we can be sure that all buckets have data up to the checkpoint.
-    yield* bucketDataInBatches({
-      storage,
-      checkpoint,
-      bucketsToFetch,
-      dataBuckets,
-      raw_data,
-      binary_data,
-      signal,
-      tracker,
-      user_id: syncParams.user_id
-    });
+    // TODO: This is not really how bucket priorities are supposed to be implemented - we might want to listen for new write
+    // checkpoints while we're serving this one too. When we're syncing low-priority buckets of this checkpoint, we could interrupt
+    // this batch and serve the high-priority buckets of the new checkpoint first?
+    bucketsToFetch.sort((a, b) => a.priority - b.priority);
+    let firstBucketInSamePriority = 0;
+    let currentPriority = bucketsToFetch[0]!.priority;
+    const lowestPriority = bucketsToFetch.at(-1)!.priority;
+
+    const bucketDataWithPriority = (endIndex?: number) => {
+      return bucketDataInBatches({
+        storage,
+        checkpoint,
+        bucketsToFetch: bucketsToFetch.slice(firstBucketInSamePriority, endIndex),
+        dataBuckets,
+        raw_data,
+        binary_data,
+        signal,
+        tracker,
+        user_id: syncParams.user_id,
+        forPriority: currentPriority !== lowestPriority ? currentPriority : undefined
+      });
+    };
+
+    for (let i = 0; i < bucketsToFetch.length; i++) {
+      if (bucketsToFetch[i].priority == currentPriority) {
+        continue;
+      }
+
+      // This incrementally updates dataBuckets with each individual bucket position.
+      // At the end of this, we can be sure that all buckets have data up to the checkpoint.
+      yield* bucketDataWithPriority(i);
+      firstBucketInSamePriority = i;
+      currentPriority = bucketsToFetch[i].priority;
+    }
```

Comment:
The logic could perhaps be more clear if you group by priority first.

`Map.groupBy` would be great for that, but would need an upgrade to Node 22 first.